### PR TITLE
add new rule no-unused-array

### LINKED ISSUE #214: Rule: Collection and array contents should be used
Implements [RSPEC-4030](https://jira.sonarsource.com/browse/RSPEC-4030)

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes #214

### CODE REVIEW

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        ]) as ts.VariableDeclaration;
+        if (varDeclaration) {
+          return !varDeclaration.initializer || Rule.isArrayLiteral(varDeclaration.initializer);
+        }
+        return true;
+      })
+      // filter out symbols with at least one read usage
+      .filter(symbol => !symbols.allUsages(symbol).some(usage => Rule.isReadUsage(usage)))
+      // raise issue
+      .forEach(symbol => walker.addFailureAtNode(Rule.getDeclaration(symbol, symbols)!, Rule.MESSAGE));
```
Comment (SonarTech) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      if (is(callExpression.expression, ts.SyntaxKind.PropertyAccessExpression)) {
+        const propertyAccess = callExpression.expression as ts.PropertyAccessExpression;
+        return propertyAccess.expression === usage.node && writingMethods.has(propertyAccess.name.text);
+      }
+    }
+
+    return false;
+  }
+
+  private static getDeclaration(symbol: ts.Symbol, symbols: SymbolTable): ts.Node | null {
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: @vilchik-elena I suggest `findDeclarationNode` here

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      .getSymbols()
+      // get only symbols storing arrays
+      .filter(symbol => Rule.hasArrayType(symbol, program.getTypeChecker(), symbols))
+      // filter out unused symbols
+      .filter(symbol => symbols.allUsages(symbol).length > 1)
+      // filter out symbols without declaration
+      .filter(symbol => {
+        const declaration = Rule.getDeclaration(symbol, symbols);
+        if (declaration) {
+          symbolDeclarations.set(symbol, declaration);
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: Consider using map to create `{ symbol, declarataion }`

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    ruleName: "no-unused-array",
+    description: "Array contents should be used",
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-4030",
+    type: "maintainability",
+    typescriptOnly: false,
+  };
+
+  private static writeArrayPatterns: ((statement: ts.ExpressionStatement, usage: Usage) => boolean)[] = [
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: I would make this all upper-case or `MESSAGE` lower-case

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    Rule.isElementWrite,
+    Rule.isVariableWrite,
+    Rule.isWritingMethodCall,
+  ];
+
+  public static MESSAGE = "Either use this array's contents or remove the array.";
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: I think it can be private. Readonly

File: sonarts-core/src/rules/noUnusedArrayRule.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    const varStatement = firstAncestor(declaration, [ts.SyntaxKind.VariableStatement]);
+    return !!varStatement && is(varStatement.getFirstToken(), ts.SyntaxKind.ExportKeyword);
+  }
+
+  private static isReadUsage(usage: Usage): boolean {
+    if (usage.is(UsageFlag.DECLARATION)) {
+      return false;
+    }
+
+    const expressionStatement = firstAncestor(usage.node, [
```
Comment (inverno) [inline sonarts-core/src/rules/noUnusedArrayRule.ts]: @vilchik-elena do you know if we can change this formatting to something that is not this ugly? This should just be `firstAncestor(usage.node, [ ts.SyntaxKind.ExpressionStatement ])` in one line.

File: sonarts-core/src/rules/noUnusedArrayRule.ts (hunk 4 above)
Comment (vilchik-elena) [reply sonarts-core/src/rules/noUnusedArrayRule.ts]: done

File: sonarts-core/src/rules/noUnusedArrayRule.ts (hunk 5 above)
Comment (vilchik-elena) [reply sonarts-core/src/rules/noUnusedArrayRule.ts]: done

**Pull request merged**